libcst/metadata/scope_provider.py

```diff
@@ -915,6 +979,33 @@ def _visit_comp_alike(
                 node.elt.visit(self)
         return False
 
+    def leave_Assign(self, original_node: cst.Assign) -> None:
+        self._leave_assignment_alike()
+
+    def leave_AnnAssign(self, original_node: cst.AnnAssign) -> None:
+        self._leave_assignment_alike()
+
+    def leave_AugAssign(self, original_node: cst.AugAssign) -> None:
+        self._leave_assignment_alike()
+
+    def leave_AsName(self, original_node: cst.AsName) -> None:
+        self._leave_assignment_alike()
+
+    def leave_CompFor(self, original_node: cst.CompFor) -> None:
+        self._leave_assignment_alike()
+
+    def leave_For(self, original_node: cst.For) -> None:
+        self._leave_assignment_alike()
+
+    def leave_NamedExpr(self, original_node: cst.NamedExpr) -> None:
+        self._leave_assignment_alike()
+
+    def leave_Parameters(self, original_node: cst.Parameters) -> None:
+        self._leave_assignment_alike()
+
+    def leave_WithItem(self, original_node: cst.WithItem) -> None:
+        self._leave_assignment_alike()
+
```

Comment:
The current way to increment assignment count require us to manual identify each leave_Node and call `self._leave_assignment_alike()` which make the code less easy to maintain and easier to miss some nodes.
Another simpler way to increment the assignment count is to simply provide a `get_and_increment_assignment_count()` method which automatically increment the index when it's called. In this way the index will be incremented whenever we create an assignment and access the assignment_count which is a simpler implementation and easier to make it right.

Reply:
This is what I started out with, but unfortunately it's not correct: `Assignment`s are created as we visit nodes, so the ordering is dependent on visitation order. For a statement like
```
b = b
```
We want the right hand side to not be able to refer to the left hand side. This means we need to increase the counter _after_ processing this statement, not just after visiting the left hand side.

Reply:
Got it. I'm a bit worry about the increasing complexity of ScopeProvider class implementation.
How about move all the assignment counter state to another Visitor class and let ScopeProvider class inherit from it? That can isolate the index counter logic and keep it simple.